Tests/Basic/SyncronizedQueueTests.swift

```diff
@@ -0,0 +1,84 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import Utility
+import Basic
+
+class SyncronizedQueueTests: XCTestCase {
+
+    func testSingleProducerConsumer() {
+        let queue = SyncronizedQueue<Int>()
+        let queueElements = Set(0..<10)
+        var consumed = Set<Int>()
+
+        let producer = Thread {
+            for element in queueElements {
+                queue.enqueue(element)
+            }
+            queue.enqueue(nil)
+        }
+
+        let consumer = Thread {
+            while let element = queue.dequeue() {
+                consumed.insert(element)
+            }
+        }
+
+        consumer.start()
+        producer.start()
+        consumer.join()
+        producer.join()
+
+       XCTAssertEqual(consumed, queueElements)
+    }
+
+    func testMultipleProducerConsumer() {
+        let queue = SyncronizedQueue<Int>()
+
+        let queueElementsOne = Set(0..<10)
+        let queueElementsTwo = Set(11..<20)
+
+        var consumed = Set<Int>()
+
+        // Create two producers.
+        let producers = [queueElementsOne, queueElementsTwo].map { queueElements in
+            return Thread {
+                for element in queueElements {
+                    queue.enqueue(element)
+                }
+                queue.enqueue(nil)
+            }
+        }
+
+        // Create two consumers.
+        let consumers = [0, 1].map { _ in
+            return Thread {
+                while let element = queue.dequeue() {
+                    consumed.insert(element)
```

Comment:
These are good tests, but I would suggest bumping the # of involved threads and iteration counts much higher (e.g., 10-100 threads and 1000 - 10000 elements) in order to truly stress test the possible race conditions. The code above should crash in some situations, so I would try first trying to adjust the parameters to get it to crash before making it safe.